File: packages/core/src/types/domain.ts
export type VerificationStatus = 'pending' | 'verified' | 'failed'
export type VerificationMethod = 'dns_txt' | 'well_known_file'

export interface Domain {
  id: string
  user_id: string
  domain: string
  verification_status: VerificationStatus
  verification_method: VerificationMethod
  verification_token: string
  verified_at: string | null
  created_at: string
  updated_at: string
}

export interface DomainVerificationInstructions {
  domain: string
  token: string
  dns_txt: {
    record_name: string
    record_type: 'TXT'
    record_value: string
  }
  well_known_file: {
    url: string
    file_path: string
    content: string
  }
}

export interface CreateDomainInput {
  domain: string
}

export interface VerifyDomainInput {
  method: VerificationMethod
}

File: apps/engine/app/schemas/domain.py
import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, field_validator
import re

_DOMAIN_RE = re.compile(
    r"^(?:[a-z0-9](?:[a-z0-9\-]{0,61}[a-z0-9])?\.)+[a-z]{2,}$",
    re.IGNORECASE,
)

VerificationStatus = Literal["pending", "verified", "failed"]
VerificationMethod = Literal["dns_txt", "well_known_file"]


class CreateDomainRequest(BaseModel):
    domain: str

    @field_validator("domain")
    @classmethod
    def validate_domain(cls, v: str) -> str:
        v = v.strip().lower()
        if not _DOMAIN_RE.match(v):
            raise ValueError("Invalid domain name (e.g. example.com)")
        return v


class VerifyDomainRequest(BaseModel):
    method: VerificationMethod = "dns_txt"


class DnsTxtInstructions(BaseModel):
    record_name: str
    record_type: str = "TXT"
    record_value: str


class WellKnownFileInstructions(BaseModel):
    url: str
    file_path: str
    content: str


class VerificationInstructions(BaseModel):
    domain: str
    token: str
    dns_txt: DnsTxtInstructions
    well_known_file: WellKnownFileInstructions


class DomainOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    user_id: uuid.UUID
    domain: str
    verification_status: VerificationStatus
    verification_method: VerificationMethod
    verification_token: str
    verified_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime


class DomainWithInstructions(BaseModel):
    domain: DomainOut
    instructions: VerificationInstructions


class DomainListResponse(BaseModel):
    domains: list[DomainOut]


class VerifyResponse(BaseModel):
    domain: DomainOut
    already_verified: bool = False
    error: Optional[str] = None
    instructions: Optional[VerificationInstructions] = None
